File: trunk/Source/2013/ThunderChicken2013/Subsystems/ProximitySensor.cpp
#include "ProximitySensor.h"
#include "NetworkCommunication/UsageReporting.h"
#include "Timer.h"
#include "WPIErrors.h"
#include "LiveWindow/LiveWindow.h"
#include <time.h>
#include <math.h>
#include "AnalogModule.h"

ProximitySensor::ProximitySensor(UINT8 analog_module,UINT32 analog_channel, SensorRange range) :
	input(analog_module, analog_channel)
{
	this->range = range;
	InitProximitySensor();
}

/**
 * Initialize the ProximitySensor.
 */
void ProximitySensor::InitProximitySensor()
{
	// Be careful not to set the average bits more than once.
	UINT32 num_average_bits = 4;
	if ( input.GetAverageBits() != num_average_bits )
	{
		input.SetAverageBits(num_average_bits);	// 2^4 = 16-sample average
		float sampleRate = 50.0 * 
			(1 << (num_average_bits));
		input.GetModule()->SetSampleRate(sampleRate);
		Wait(0.2);
	}
	LiveWindow::GetInstance()->AddSensor("Tilter","Ultrasonic", this);
}

/**
 * Delete the ProximitySensor.
 */
ProximitySensor::~ProximitySensor()
{
}

/**
 * Return the actual distance in millimeters.
 * 
 * @return the current distance in millimeters.
 */

float ProximitySensor::GetDistanceMM( void )
{
	float voltage = input.GetAverageVoltage();
	
	// Transform voltage to distance.  Exponential values
	// were derived from the datasheets:
	//
	// GP2D120XJ00F (Short Range [30cm]):  https://www.sparkfun.com/datasheets/Sensors/Infrared/GP2D120XJ00F_SS.pdf
	// GP2Y0A21YK (Medium Range [80cm]):  http://www.sparkfun.com/datasheets/Components/GP2Y0A21YK.pdf
	// GP2Y0A02YK0F (Long Range [150cm]):  https://www.sparkfun.com/datasheets/Sensors/Infrared/gp2y0a02yk_e.pdf
	
	float distance_mm;
	switch ( range )
	{
	case kShortRange:
	
		distance_mm = 126.27 * pow(voltage, -1.102);
		break;
		
	case kMediumRange:
		distance_mm = 267.1 * pow(voltage, -1.23);
		break;
		
	case kLongRange:
		distance_mm = 0.0;	// TODO
		break;
	}
	
	return distance_mm;
}

/**
 * Get the angle in degrees for the PIDSource base object.
 * 
 * @return The angle in degrees.
 */
double ProximitySensor::PIDGet()
{
	return GetDistanceMM();
}

void ProximitySensor::UpdateTable() {
	if (m_table != NULL) {
		m_table->PutNumber("Value", GetDistanceMM());
	}
}

void ProximitySensor::StartLiveWindowMode() {
	
}

void ProximitySensor::StopLiveWindowMode() {
	
}

std::string ProximitySensor::GetSmartDashboardType() {
	return "Ultrasonic";
}

void ProximitySensor::InitTable(ITable *subTable) {
	m_table = subTable;
	UpdateTable();
}

ITable * ProximitySensor::GetTable() {
	return m_table;
}

File: Source/2014/CompetitionRobot/Robot/src/org/usfirst/frc2465/Robot/subsystems/ProximitySensor.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package org.usfirst.frc2465.Robot.subsystems;

import edu.wpi.first.wpilibj.AnalogChannel;
import edu.wpi.first.wpilibj.PIDSource;
import edu.wpi.first.wpilibj.SensorBase;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.livewindow.LiveWindowSendable;
import edu.wpi.first.wpilibj.tables.ITable;
import com.sun.squawk.util.MathUtils;

/**
 *
 * @author Scott
 */

public class ProximitySensor extends SensorBase implements PIDSource, LiveWindowSendable {

    public final int kShortRange    = 0; 
    public final int kMediumRange   = 1; 
    public final int kLongRange     = 2; 

    int sensor_range;
    AnalogChannel   analog_channel;
    ITable table;
    
    public ProximitySensor(final int slot, final int channel, int sensor_range) {
        this.sensor_range = sensor_range;
        analog_channel = new AnalogChannel(slot, channel);
        initProximitySensor();
    }
    
    /**
     * Initialize the ProximitySensor.
     */
    protected void initProximitySensor()
    {
        // Be careful not to set the average bits more than once.
        int num_average_bits = 4;
        if ( analog_channel.getAverageBits() != num_average_bits )
        {
            analog_channel.setAverageBits(num_average_bits); // 2^4 = 16-sample average
            float sampleRate = (float) (50.0 * 
                    (1 << (num_average_bits)));
            analog_channel.getModule().setSampleRate(sampleRate);
            Timer.delay(0.2);
        }
    }

    /**
     * Return the actual distance in millimeters.
     * 
     * @return the current distance in millimeters.
     */

    public double getDistanceMM()
    {
        double voltage = analog_channel.getAverageVoltage();
        
        // Transform voltage to distance.  Exponential values
        // were derived from the datasheets:
        //
        // GP2D120XJ00F (Short Range [30cm]):  https://www.sparkfun.com/datasheets/Sensors/Infrared/GP2D120XJ00F_SS.pdf
        // GP2Y0A21YK (Medium Range [80cm]):  http://www.sparkfun.com/datasheets/Components/GP2Y0A21YK.pdf
        // GP2Y0A02YK0F (Long Range [150cm]):  https://www.sparkfun.com/datasheets/Sensors/Infrared/gp2y0a02yk_e.pdf
        
        double distance_mm = 0.0;
        switch ( sensor_range )
        {
        case kShortRange:
        
                distance_mm = 126.27 * MathUtils.pow(voltage, -1.102);
                break;
                
        case kMediumRange:
                distance_mm = 267.1 * MathUtils.pow(voltage, -1.23);
                break;
                
        case kLongRange:
                distance_mm = 61.681 * MathUtils.pow(voltage, -1.133);
                break;
        }
        
        return distance_mm;
    }
    
    
    public double pidGet() {
        return getDistanceMM();
    }

    public void updateTable() {
        if (table != null) {
            table.putNumber("Value", getDistanceMM());
        }
    }

    public void startLiveWindowMode() {
    }

    public void stopLiveWindowMode() {
    }

    public void initTable(ITable itable) {
        table = itable;
        updateTable();
    }

    public ITable getTable() {
        return table;
    }

    public String getSmartDashboardType() {
        return "Ultrasonic";
    }
    
}
